File: java/Llaves.java
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.security.spec.NamedParameterSpec;
import java.util.Arrays;
import javax.crypto.KeyAgreement;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

final class Llaves {
    record SessionKeys(byte[] z, byte[] salt, byte[] prk, byte[] kPermBase, byte[] kKernelBase, byte[] kMac, byte[] kSeed) {}
    record RoundKeys(byte[] kPerm, byte[] kKernel, int permIndex, int[] moore1, int[] moore2) {}

    static byte[] generateSalt() {
        byte[] salt = new byte[Common.SALT_BYTES];
        new SecureRandom().nextBytes(salt);
        return salt;
    }

    static byte[] deriveSharedSecret() throws Exception {
        KeyPairGenerator kpg = KeyPairGenerator.getInstance("X25519");
        kpg.initialize(new NamedParameterSpec("X25519"));
        KeyPair a = kpg.generateKeyPair();
        KeyPair b = kpg.generateKeyPair();
        KeyAgreement ka = KeyAgreement.getInstance("X25519");
        ka.init(a.getPrivate());
        ka.doPhase(b.getPublic(), true);
        byte[] za = ka.generateSecret();
        KeyAgreement kb = KeyAgreement.getInstance("X25519");
        kb.init(b.getPrivate());
        kb.doPhase(a.getPublic(), true);
        byte[] zb = kb.generateSecret();
        if (!Arrays.equals(za, zb)) throw new IllegalStateException("No coinciden los secretos X25519.");
        return za;
    }

    static SessionKeys deriveSession(byte[] z, byte[] salt, Common.RcaContext ctx) throws Exception {
        byte[] prk = hkdfExtract(salt, z);
        return new SessionKeys(
            z.clone(),
            salt.clone(),
            prk,
            hkdfExpand(prk, buildInfo("RCA|perm|v2", ctx), 32),
            hkdfExpand(prk, buildInfo("RCA|kernel|v2", ctx), 32),
            hkdfExpand(prk, buildInfo("RCA|mac|v2", ctx), 32),
            hkdfExpand(prk, buildInfo("RCA|drbg|v2", ctx), 32)
        );
    }

    static RoundKeys deriveRound(SessionKeys session, Common.RcaContext ctx, int round) throws Exception {
        byte[] kPerm = hkdfExpand(session.prk(), buildRoundInfo("RCA|perm|v2|round=", round, ctx), 32);
        byte[] kKernel = hkdfExpand(session.prk(), buildRoundInfo("RCA|kernel|v2|round=", round, ctx), 32);
        byte permIndex = (byte)(MessageDigest.getInstance("SHA-256").digest(kPerm)[0] & 0x7F);
        return new RoundKeys(kPerm, kKernel, permIndex & 0xFF, deriveMoore(kKernel, (byte)1), deriveMoore(kKernel, (byte)2));
    }

    static Common.BoundaryConfig boundaryFromRound(RoundKeys rk) {
        return new Common.BoundaryConfig(
            Common.BoundaryMode.values()[(rk.kKernel()[0] & 0xFF) % 3],
            Common.BoundaryMode.values()[(rk.kKernel()[1] & 0xFF) % 3],
            Common.BoundaryMode.values()[(rk.kKernel()[2] & 0xFF) % 3],
            Common.BoundaryMode.values()[(rk.kKernel()[3] & 0xFF) % 3]
        );
    }

    private static byte[] hkdfExtract(byte[] salt, byte[] ikm) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec((salt == null || salt.length == 0) ? new byte[32] : salt, "HmacSHA256"));
        return mac.doFinal(ikm);
    }

    private static byte[] hkdfExpand(byte[] prk, byte[] info, int length) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec(prk, "HmacSHA256"));
        byte[] okm = new byte[length];
        byte[] t = new byte[0];
        int off = 0;
        byte counter = 1;
        while (off < length) {
            ByteArrayOutputStream block = new ByteArrayOutputStream();
            block.write(t);
            block.write(info);
            block.write(counter++);
            t = mac.doFinal(block.toByteArray());
            int copy = Math.min(t.length, length - off);
            System.arraycopy(t, 0, okm, off, copy);
            off += copy;
        }
        return okm;
    }

    private static byte[] buildInfo(String label, Common.RcaContext ctx) throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        out.write(label.getBytes(StandardCharsets.US_ASCII));
        out.write("RCA|ctx|v2".getBytes(StandardCharsets.US_ASCII));
        writeLe32(out, ctx.height());
        writeLe32(out, ctx.width());
        writeLe16(out, ctx.channels());
        writeLe16(out, ctx.rounds());
        return out.toByteArray();
    }

    private static byte[] buildRoundInfo(String label, int round, Common.RcaContext ctx) throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        out.write(label.getBytes(StandardCharsets.US_ASCII));
        writeLe32(out, round);
        out.write("RCA|ctx|v2".getBytes(StandardCharsets.US_ASCII));
        writeLe32(out, ctx.height());
        writeLe32(out, ctx.width());
        writeLe16(out, ctx.channels());
        writeLe16(out, ctx.rounds());
        return out.toByteArray();
    }

    private static int[] deriveMoore(byte[] kKernel, byte tag) throws Exception {
        byte[] material = Arrays.copyOf(kKernel, 33);
        material[32] = tag;
        byte[] digest = MessageDigest.getInstance("SHA-256").digest(material);
        int[] out = new int[8];
        for (int i = 0; i < 8; i++) {
            int w = Short.toUnsignedInt(ByteBuffer.wrap(digest, i * 2, 2).order(ByteOrder.LITTLE_ENDIAN).getShort());
            out[i] = (w % 256) + 1;
        }
        return out;
    }

    private static void writeLe16(OutputStream out, int value) throws IOException {
        out.write(value & 0xFF);
        out.write((value >>> 8) & 0xFF);
    }

    private static void writeLe32(OutputStream out, int value) throws IOException {
        out.write(value & 0xFF);
        out.write((value >>> 8) & 0xFF);
        out.write((value >>> 16) & 0xFF);
        out.write((value >>> 24) & 0xFF);
    }
}

File: cs/Llaves.cs
using System;
using System.Buffers.Binary;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Security;

namespace TTCrypto;

public sealed class SessionKeys
{
    public byte[] Z { get; init; } = new byte[CryptoConsts.KeyBytes];
    public byte[] Salt { get; init; } = new byte[CryptoConsts.SaltBytes];
    public byte[] Prk { get; init; } = new byte[CryptoConsts.KeyBytes];
    public byte[] KPermBase { get; init; } = new byte[CryptoConsts.KeyBytes];
    public byte[] KKernelBase { get; init; } = new byte[CryptoConsts.KeyBytes];
    public byte[] KMac { get; init; } = new byte[CryptoConsts.KeyBytes];
    public byte[] KSeed { get; init; } = new byte[CryptoConsts.KeyBytes];
}

public sealed class RoundKeys
{
    public byte[] KPerm { get; init; } = new byte[CryptoConsts.KeyBytes];
    public byte[] KKernel { get; init; } = new byte[CryptoConsts.KeyBytes];
    public byte PermIndex { get; init; }
    public ushort[][][] MooreByChannel { get; init; } =
    [
        [new ushort[8], new ushort[8]],
        [new ushort[8], new ushort[8]],
        [new ushort[8], new ushort[8]]
    ];
}

public static class Keys
{
    public static byte[] GenerateSalt()
    {
        var salt = new byte[CryptoConsts.SaltBytes];
        RandomNumberGenerator.Fill(salt);
        return salt;
    }

    public static (byte[] PrivateKey, byte[] PublicKey) GenerateX25519Pair()
    {
        var privateKey = new X25519PrivateKeyParameters(new SecureRandom());
        var publicKey = privateKey.GeneratePublicKey();
        var sk = new byte[X25519PrivateKeyParameters.SecretSize];
        var pk = new byte[X25519PublicKeyParameters.KeySize];
        privateKey.Encode(sk, 0);
        publicKey.Encode(pk, 0);
        return (sk, pk);
    }

    public static byte[] SharedSecretFromPkcs8(byte[] privateKeyPkcs8, byte[] publicKeySpki)
    {
        if (privateKeyPkcs8.Length != X25519PrivateKeyParameters.SecretSize)
            throw new ArgumentException("La llave privada X25519 debe tener 32 bytes.", nameof(privateKeyPkcs8));
        if (publicKeySpki.Length != X25519PublicKeyParameters.KeySize)
            throw new ArgumentException("La llave publica X25519 debe tener 32 bytes.", nameof(publicKeySpki));

        var privateKey = new X25519PrivateKeyParameters(privateKeyPkcs8, 0);
        var publicKey = new X25519PublicKeyParameters(publicKeySpki, 0);
        var shared = new byte[X25519PrivateKeyParameters.SecretSize];
        privateKey.GenerateSecret(publicKey, shared, 0);
        return shared;
    }

    public static SessionKeys DeriveSession(byte[] z, byte[] salt, RcaContext ctx)
    {
        var prk = HkdfExtract(salt, z);
        return new SessionKeys
        {
            Z = z.ToArray(),
            Salt = salt.ToArray(),
            Prk = prk,
            KPermBase = HkdfExpand(prk, BuildInfo("RCA|perm|v2", ctx), 32),
            KKernelBase = HkdfExpand(prk, BuildInfo("RCA|kernel|v2", ctx), 32),
            KMac = HkdfExpand(prk, BuildInfo("RCA|mac|v2", ctx), 32),
            KSeed = HkdfExpand(prk, BuildInfo("RCA|drbg|v2", ctx), 32)
        };
    }

    public static RoundKeys DeriveRound(SessionKeys session, RcaContext ctx, uint roundIndex)
    {
        var kPerm = HkdfExpand(session.Prk, BuildRoundInfo("RCA|perm|v2|round=", roundIndex, ctx), 32);
        var kKernel = HkdfExpand(session.Prk, BuildRoundInfo("RCA|kernel|v2|round=", roundIndex, ctx), 32);
        var mooreByChannel = new ushort[3][][];
        for (var ch = 0; ch < 3; ch++)
        {
            mooreByChannel[ch] = new ushort[2][];
            for (var k = 0; k < 2; k++)
                mooreByChannel[ch][k] = DeriveMoore(kKernel, (byte)(1 + ch * 2 + k));
        }
        return new RoundKeys
        {
            KPerm = kPerm,
            KKernel = kKernel,
            PermIndex = (byte)(SHA256.HashData(kPerm)[0] & 0x7F),
            MooreByChannel = mooreByChannel
        };
    }

    private static ushort[] DeriveMoore(byte[] kKernel, byte tag)
    {
        var material = new byte[33];
        Array.Copy(kKernel, material, 32);
        material[32] = tag;
        var digest = SHA256.HashData(material);
        var coeffs = new ushort[8];
        for (var i = 0; i < 8; i++)
        {
            var w = BinaryPrimitives.ReadUInt16LittleEndian(digest.AsSpan(i * 2, 2));
            coeffs[i] = (ushort)((w % 256) + 1);
        }
        return coeffs;
    }

    private static byte[] BuildInfo(string label, RcaContext ctx)
    {
        using var ms = new MemoryStream();
        ms.Write(Encoding.ASCII.GetBytes(label));
        WriteInfoSuite(ms, ctx);
        return ms.ToArray();
    }

    private static byte[] BuildRoundInfo(string label, uint roundIndex, RcaContext ctx)
    {
        using var ms = new MemoryStream();
        ms.Write(Encoding.ASCII.GetBytes(label));
        Span<byte> u32 = stackalloc byte[4];
        BinaryPrimitives.WriteUInt32LittleEndian(u32, roundIndex);
        ms.Write(u32);
        WriteInfoSuite(ms, ctx);
        return ms.ToArray();
    }

    private static void WriteInfoSuite(Stream stream, RcaContext ctx)
    {
        stream.Write(Encoding.ASCII.GetBytes("RCA|ctx|v2"));
        Span<byte> u32 = stackalloc byte[4];
        Span<byte> u16 = stackalloc byte[2];
        BinaryPrimitives.WriteUInt32LittleEndian(u32, ctx.Height);
        stream.Write(u32);
        BinaryPrimitives.WriteUInt32LittleEndian(u32, ctx.Width);
        stream.Write(u32);
        BinaryPrimitives.WriteUInt16LittleEndian(u16, ctx.Channels);
        stream.Write(u16);
        BinaryPrimitives.WriteUInt16LittleEndian(u16, ctx.Rounds);
        stream.Write(u16);
    }

    private static byte[] HkdfExtract(byte[] salt, byte[] ikm)
    {
        using var hmac = new HMACSHA256(salt.Length == 0 ? new byte[32] : salt);
        return hmac.ComputeHash(ikm);
    }

    private static byte[] HkdfExpand(byte[] prk, byte[] info, int length)
    {
        using var hmac = new HMACSHA256(prk);
        var okm = new byte[length];
        var t = Array.Empty<byte>();
        var offset = 0;
        byte counter = 1;
        while (offset < length)
        {
            var block = new byte[t.Length + info.Length + 1];
            Buffer.BlockCopy(t, 0, block, 0, t.Length);
            Buffer.BlockCopy(info, 0, block, t.Length, info.Length);
            block[^1] = counter++;
            t = hmac.ComputeHash(block);
            var copy = Math.Min(t.Length, length - offset);
            Buffer.BlockCopy(t, 0, okm, offset, copy);
            offset += copy;
        }
        return okm;
    }
}
